### PR TITLE
Fix crash bug when a child is reporting a value when using MultiprocessParallelUpdater

### LINKED ISSUE #3389: `MultiprocessParallelUpdater` does not add child links as observers to reporter.
Unlike `Trainer`, `MultiprocessParallelUpdater` does not add child links as observers to reporter.
Since the Chain’s namedlinks() are not added to the reporter, this means that when running multi GPU training, the reporter cannot find the observer, which leads to Chainer crashing.
Basically, this is due to the discrepancy in the following lines:
https://github.com/chainer/chainer/blob/master/chainer/training/updaters/multiprocess_parallel_updater.py#L39
https://github.com/chainer/chainer/blob/master/chainer/training/trainer.py#L147-L149

A simple fix would be to add `self.reporter.add_observers('main', self.model.namedlinks(skipself=True))`
to `chainer.training.updaters.multiprocess_parallel_updater._Worker`'s `setup` function.
Might create a PR later, but just mentioning this for now.

### PR DESCRIPTION
**Pull request by tkerola**
Fixes #3389 
Without the suggested change, Chainer will crash with error message `KeyError: 'Given observer is not registered to the reporter.'` when reporting values from a child that resides on a different GPU.

Being able to report values from a child Chain is useful for monitoring certain values of the network.

### CODE REVIEW

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by tkerola** [comment]
Hmm, the AppVeyor build seems to have failed due to `Build execution time has reached the maximum allowed time for your plan (60 minutes).`. I am not quite sure why. The tests work on my machine (although I only tested on Ubuntu 14.04).

**Comment by niboshi** [comment]
This problem is caused by `test_multiprocess_iterator` (#3383). It's disabled in the lastest master branch. Please rebase to master. I'm sorry for inconvenience.

**Comment by tkerola** [comment]
It seems like the TravisCI build was cancelled for some reason. I cannot see any error in the log, so I wonder why this happened. Is it possible to restart the build?

**Comment by niboshi** [comment]
Restarted.

**Comment by tkerola** [comment]
Is there any issue left with this PR that needs to be resolved?

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
Jenkins, test this please

File: tests/chainer_tests/training_tests/updaters_tests/test_multiprocess_parallel_updater.py [hunk 1]
```diff
@@ -180,4 +183,73 @@ def test_update_uses_raw_array(self):
             self.assertEqual(model.call_called, 1)
 
 
+class SimpleNetChild(chainer.Chain):
+
+    def __init__(self):
+        super(SimpleNetChild, self).__init__(
```
Comment (okuta) [inline tests/chainer_tests/training_tests/updaters_tests/test_multiprocess_parallel_updater.py]: Please use new style.
```python
super(SimpleNetChild, self).__init__()
with self.init_scope():
    self.conv = chainer.links.Convolution2D(2, 2, 3)
```

File: tests/chainer_tests/training_tests/updaters_tests/test_multiprocess_parallel_updater.py [hunk 2]
```diff
@@ -180,4 +183,73 @@ def test_update_uses_raw_array(self):
...
+        chainer.reporter.report({
+            'h_max': chainer.functions.math.minmax.max(h)}, self)
+
+        return h
+
+
+class SimpleNetChildReporter(chainer.Chain):
+
+    def __init__(self):
+        super(SimpleNetChildReporter, self).__init__(
```
Comment (okuta) [inline tests/chainer_tests/training_tests/updaters_tests/test_multiprocess_parallel_updater.py]: ditto.

**Comment by okuta** [comment]
@tkerola Sorry late response.
I checked your PR.

**Comment by tkerola** [comment]
Thank you for your review. I updated the PR according to your comments.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
I found a critical bug on Python3 environment. I will send PR. Please wait.

**Comment by okuta** [comment]
@tkerola  Please merge [this PR](https://github.com/tkerola/chainer/pull/1) to fix test.

**Comment by tkerola** [comment]
@okuta Ok, I merged it. Thank you very much for the PR!

**Comment by okuta** [comment]
Thanks!
jenkins, test this please.

File: tests/chainer_tests/training_tests/updaters_tests/test_multiprocess_parallel_updater.py [hunk 3]
```diff
@@ -180,4 +187,98 @@ def test_update_uses_raw_array(self):
...
+    iters = [chainer.iterators.SerialIterator(i, batch_size) for i in
+             chainer.datasets.split_dataset_n_random(
+                 dataset, len(devices))]
+    optimizer = chainer.optimizers.SGD(lr=1.0)
+    optimizer.setup(model)
+    updater = mpu.MultiprocessParallelUpdater(
+        iters, optimizer, devices=devices)
+    trainer = trainer.Trainer(updater, (1, 'iteration'), '/tmp')
+    trainer.run()
+    assert model.call_called == 0
```
Comment (okuta) [inline tests/chainer_tests/training_tests/updaters_tests/test_multiprocess_parallel_updater.py]: Sorry, this is my mistake. Please fix to `1`.

**Comment by okuta** [comment]
Thanks!
jenkins, test this please.

**Pull request merged**